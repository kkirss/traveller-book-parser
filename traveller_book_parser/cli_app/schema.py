import json
import logging
import pathlib
from typing import Optional

from pydantic import BaseModel
import typer

from traveller_book_parser.books.book_description import BookDescription
from traveller_book_parser.settings import SETTINGS
from traveller_book_parser.traveller_models.trav_object import TravObjectRoot
from traveller_book_parser.traveller_models.traveller_database import TravellerDatabase
from traveller_book_parser.utils import ensure_folder

logger = logging.getLogger(__name__)

schema_app = typer.Typer()


def dump_model_schema(
    path: Optional[pathlib.Path],
    cls: type[BaseModel],
    name: str,
):
    """Dump the JSON schema of a model."""
    if not cls.__doc__:
        logger.warning("Model '%s' has no docstring.", cls.__name__)

    if path is None:
        path = SETTINGS.default_schema_output_path / f"{name}.json"
    elif path.is_dir():
        path = path / f"{name}.json"

    ensure_folder(path.parent)

    with path.open("w") as f:
        json.dump(cls.model_json_schema(), f, indent=2)


def book_description_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'BookDescription' model."""
    dump_model_schema(path, BookDescription, "BookDescription")


schema_app.command("book_description")(book_description_schema_cli)
schema_app.command("BookDescription")(book_description_schema_cli)


def database_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'TravellerDatabase' model."""
    dump_model_schema(path, TravellerDatabase, "TravellerDatabase")


schema_app.command("traveller_database")(database_schema_cli)
schema_app.command("TravellerDatabase")(database_schema_cli)


def trav_object_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump the JSON schema of the 'TravObject' model."""
    dump_model_schema(path, TravObjectRoot, "TravObject")


schema_app.command("trav_object")(trav_object_schema_cli)
schema_app.command("TravObject")(trav_object_schema_cli)


def all_schema_cli(path: Optional[pathlib.Path] = None):
    """Dump all JSON schema files."""
    book_description_schema_cli(path)
    database_schema_cli(path)
    trav_object_schema_cli(path)


schema_app.command("all")(all_schema_cli)
