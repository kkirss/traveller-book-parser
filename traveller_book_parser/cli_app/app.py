import typer

from .parse import parse_all_books_cli, parse_book_cli

app = typer.Typer(
    help="Parse Traveller books into other formats.",
    no_args_is_help=True,
)


app.command("parse-book", no_args_is_help=True)(parse_book_cli)
app.command("parse")(parse_all_books_cli)


def run_cli():
    """Run CLI."""
    app()
