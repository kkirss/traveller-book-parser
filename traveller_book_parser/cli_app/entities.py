from traveller_book_parser.traveller_models.entity import Entity


def print_entities(entities: list[Entity]):
    for entity in entities:
        print(entity)
