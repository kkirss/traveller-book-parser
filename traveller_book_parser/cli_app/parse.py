from pathlib import Path
from typing import Optional

import typer

from traveller_book_parser.books.parse_books import parse_all_books, parse_book
from traveller_book_parser.traveller_database.output import outputted_database
from traveller_book_parser.traveller_database.output_format.output_format import (
    OutputFormat,
)
from traveller_book_parser.traveller_database.output_io.output_type import OutputType
from traveller_book_parser.traveller_database.output_options import (
    DatabaseOutputOptions,
)

parse_app = typer.Typer(no_args_is_help=True)


def parse_book_cli(
    book_code_name: str,
    output_format: OutputFormat = OutputFormat.json,
    output_type: OutputType = OutputType.stdout,
    output_path: Optional[Path] = None,
):
    """Parse a book and print processed objects."""
    options = DatabaseOutputOptions(
        output_format=output_format,
        output_type=output_type,
        output_path=output_path,
    )
    with outputted_database(options) as database:
        parse_book(database, book_code_name)


parse_app.command("book")(parse_book_cli)


def parse_all_books_cli(
    output_format: OutputFormat = OutputFormat.json,
    output_type: OutputType = OutputType.stdout,
    output_path: Optional[Path] = None,
):
    """Parse all books and print processed objects."""
    options = DatabaseOutputOptions(
        output_format=output_format,
        output_type=output_type,
        output_path=output_path,
    )
    with outputted_database(options) as database:
        parse_all_books(database)


parse_app.command("all")(parse_all_books_cli)
