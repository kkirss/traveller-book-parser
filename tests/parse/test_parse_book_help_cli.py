from syrupy.assertion import SnapshotAssertion
from typer.testing import CliRunner

from traveller_book_parser.cli_app.app import app


def test_parse_book_help_cli(snapshot: SnapshotAssertion, cli_runner: CliRunner):
    # Arrange
    command = ["parse", "book", "--help"]

    # Act
    result = cli_runner.invoke(app, [*command])

    # Assert
    assert result.exit_code == 0, result.output
    assert result.output == snapshot
