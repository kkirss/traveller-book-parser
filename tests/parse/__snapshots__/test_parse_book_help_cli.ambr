# serializer version: 1
# name: test_parse_book_help_cli
  '''
                                                                                 
   Usage: root parse book [OPTIONS] BOOK_CODE_NAME                               
                                                                                 
   Parse a book and print processed objects.                                     
                                                                                 
  ┌─ Arguments ─────────────────────────────────────────────────────────────────┐
  │ *    book_code_name      TEXT  [default: None] [required]                   │
  └─────────────────────────────────────────────────────────────────────────────┘
  ┌─ Options ───────────────────────────────────────────────────────────────────┐
  │ --output-format        [json|none]    [default: json]                       │
  │ --output-type          [stdout|file]  [default: stdout]                     │
  │ --output-path          PATH           [default: None]                       │
  │ --help                                Show this message and exit.           │
  └─────────────────────────────────────────────────────────────────────────────┘
  
  
  '''
# ---
